text/0112-server-engine.md

```diff
@@ -0,0 +1,558 @@
+
+**No access to web platform APIs on the server:** When running in the server environment, LWC will not polyfill web platform-specific APIs. Because of this, accessing any of those APIs as the component renders on the server will result in a runtime exception. For example, if a server-rendered component wants to attach event listeners to the `document` when it is rendered on the client, it needs to check first if the `document` object is present in the current runtime environment.
+
+```js
+import { LightningElement } from 'lwc';
+
+// Will evaluate to true when running in a browser, otherwise it will evaluate to false.
+const isBrowser = typeof document !== 'undefined';
+
+export default class App extends LightningElement {
+    connectedCallback() {
+        if (isBrowser) {
+            document.addEventListener('click', this.handleDocumentClick);
+        }
+    }
+
+    disconnectedCallback() {
+        if (isBrowser) {
+            document.removeEventListener('click', this.handleDocumentClick);
+        }
+    }
+
+    handleDocumentClick() { /* ... */ }
+}
+```
+
+**The entire component tree will be synchronously rendered in a single pass:** This behavior matches the current behavior of the LWC engine. Connecting an LWC component to a document triggers a synchronous rendering cycle. This means that reactivity is unnecessary on the server. Disabling the reactive membrane on the server will also improve the overall SSR performance by not creating unnecessary JavaScript `Proxies`.
+
+**No asynchronous operations allowed:** This also means that if a component needs to do an asynchronous operation to fetch data, it will not be possible to do so when rendered on the server. All the asynchronous data dependencies for a component subtree needs to be retrieved prior to rendering the component and needs to be passed via public properties for it to be rendered. Without adding new primitives to the LWC framework, updating the state of a component asynchronously would violate the previous constraints. Other popular UI frameworks are currently working on supporting asynchronous rendering and integrating it into their current SSR solution, but it is a complex feature to implement.
+
+**The `renderedCallback` lifecycle hook will not execute on the server:** When running in a browser, this hook is the first life cycle hook which gives the component author access to the rendered DOM elements. If the component were to attempt to access those APIs on the server it would result in a runtime error since the DOM APIs are not mocked.
+
+**Wire adapters will not be invoked:** The Wire protocol emits a stream of data to a component. The current protocol doesn't define a way today to indicate that the stream is done emitting new data. Because of this, the first version of SSR will not invoke any wire adapter. The protocol will need to be changed and new primitives will need to be added to LWC to make wire adapters compatible with SSR.
```

Comment:
We should consider adding an option to capture a one time snapshot of the stream. Disabling wire adapters during SSR  eliminates the ability to use wire protocol to inject initial state including localized text in the SSR output.

Reply:
As of today, the invocation timing of the `update` callback is different if the wire adapter [has dynamic parameters or not](https://github.com/salesforce/lwc/blob/e1d7efa3e3d20881f8f11caba623492cd6e6addb/packages/%40lwc/engine-core/src/framework/wiring.ts#L336-L340). This difference makes it impossible to invoke wire adapters in a reliable fashion.

Changing the current invocation timing is an observable change that might result in breaking existing customers on the platform. We decided to set wire adapters aside from the initial SSR implementation but it is something that we will reconsider in the future.

Reply:
Did you thought about creating a new hook that fetches data on the server, before serving the component? Nuxt did that, it is called `fetch`. I believe it would be something really great for lwc. With a `fetch` method, components could be served straight away with the data developers need from Salesforce. Pages would be rendered so much faster, not just because the component was rendered on the server, but also because components would no longer need to wait for the wiredMethod to be added to the stack call on the client, which can be slow depending on the specs of the computer, and how many other wiredMethods were added to the stack.

Reply:
Perhaps this can enabled as an optional configuration param when we call `renderLwcToString` or something like that. By default wire adapters can be skipped, but users can opt-in if their use case is deterministic like in my case, I don't connect wire adapters to network requests, they are simply connected to redux style stores to decouple state and UI.

Reply:
Perhaps a reasonable compromise would be to support Wire adapters without dynamic parameters during SSR, which would cover several use-cases (including my own).

Reply:
IMO, we should make wire execution predictable (both on the client and on the server). It would be surprising that for a developer to discover that a wire stops getting executed in SSR after adding a dynamic param. 

Over the last 6 months, the LWC team has been investigating how we could roll out potentially risky changes like this one (changing the invocation timing of wires) without breaking existing customers in production. I hope that we could leverage one of the approaches we are discussing to solve this issue and enable wire adapter execution for SSR. 

/cc @nolanlawson